scheduler/scheduler.go

```diff
@@ -220,10 +221,139 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
 		"span":      time.Since(startTime).String(),
 		"duration":  time.Since(startTime),
 	})
-
 	return hostsSpawnedPerDistro, nil
 }
 
+// generateHostOptions generates host options based on what kind of host it is:
+// regular host or container
+func generateHostOptions(d distro.Distro) (cloud.HostOptions, error) {
+	if d.MaxContainers > 0 {
+		parent, err := findAvailableParent(d)
+		if err != nil {
+			err = errors.Wrap(err, "Could not find available parent host")
+			return cloud.HostOptions{}, err
+		}
+		hostOptions := cloud.HostOptions{
+			ParentID: parent.Id,
+			UserName: evergreen.User,
+		}
+		return hostOptions, nil
+	}
+
+	hostOptions := cloud.HostOptions{
+		UserName: evergreen.User,
+	}
+	return hostOptions, nil
+}
+
+// generateParentHostOptions generates host options for a parent host
+func generateParentHostOptions() cloud.HostOptions {
+	return cloud.HostOptions{
+		HasContainers: true,
+		UserName:      evergreen.User,
+	}
+}
+
+// FindAvailableParent finds a parent host that can accommodate container,
+// packing on parent that has task with longest expected finish time
+func findAvailableParent(d distro.Distro) (host.Host, error) {
+	allParents, err := host.FindAllRunningParentsOrdered()
+	if err != nil {
+		return host.Host{}, errors.Wrap(err, "Could not find running parent hosts")
+	}
+
+	// parents come in sorted order from soonest to latest expected finish time
+	for i := len(allParents) - 1; i >= 0; i-- {
+		parent := allParents[i]
+		currentContainers, err := parent.GetContainers()
+		if err != nil {
+			return host.Host{}, errors.Wrapf(err, "Could not find containers for parent %s", parent.Id)
+		}
+		if len(currentContainers) < d.MaxContainers {
+			return parent, nil
+		}
+	}
+	return host.Host{}, errors.New("No available parent found for container")
+}
+
+// numNewParentsNeeded returns the number of additional parents needed to
+// accommodate new containers
+func numNewParentsNeeded(numCurrentParents, numContainersNeeded, numExistingContainers int, d distro.Distro) int {
+	if numCurrentParents*d.MaxContainers < numExistingContainers+numContainersNeeded {
```

Comment:
This calculation definitely works, but I wonder if we want to be more conservative in scenarios where we currently have low utilization but a sudden surge in traffic.

For example if you currently have 2 parents (capacity 100) each 50% utilized, and then we decide that we need an additional 201 containers, we will start 3 new parents. That brings us to 5 parents running 301 containers. It probably won't make a difference in practice but worth considering